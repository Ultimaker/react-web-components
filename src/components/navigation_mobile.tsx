import * as React from 'react';
import { NavLink } from 'react-router-dom';

// components
import Button from './button';
import ProfileImage from './profile_image';
import SlideOutContainer from './slide_out_container';
import SlideInPanel from './slide_in_panel';
import SubNavigationMobileMenu from './sub_navigation_mobile_menu';

// components
import { NavRoute } from './Navigation';

export interface NavigationMobileProps {
    /** Details for the visible nav items */
    visibleNavLinks: NavRoute[];
    /** URL for the account portal */
    manageAccountURL?: string;
    /** Called when sign out is clicked */
    onSignOutClickHandler?: () => void;
    /** Called when the sign in button is clicked */
    onSignInClickHandler?: () => void;
    /** The label for account navigation on mobile */
    accountButtonText?: string;
    /** The label for sign out on mobile */
    signOutButtonText?: string;
    /** The label for sign in */
    signInButtonText?: string;
    /** The username of the user to be shown in the mobile menu */
    accountDisplayName?: string;
    /** The URL to the user's account page */
    accountImageURL?: string;
    /** Whether the user is signed out */
    signedOut?: boolean;
    /**
     * Used to toggle the visibility of the mobile navigation menu.
     * Passed in by the header component.
     */
    showMobileMenu?: boolean;
    /**
     * Called when the mobile navigation is closed.
     * Passed in by the header component.
     */
    onCloseMobileMenuHandler?: () => void;
    /** Whether the user profile should be displayed in the mobile menu */
    showMobileAccountNav?: boolean;
}

export class NavigationMobile extends React.Component<NavigationMobileProps, {}> {
    constructor(props) {
        super(props);

        this._onCloseMobileMenuHandler = this._onCloseMobileMenuHandler.bind(this);
    }

    private _onCloseMobileMenuHandler(): void {
        const { onCloseMobileMenuHandler } = this.props;

        if (onCloseMobileMenuHandler) {
            onCloseMobileMenuHandler();
        }
    }

    private _renderAccountNav(): JSX.Element {
        const {
            onSignOutClickHandler, onSignInClickHandler, manageAccountURL, signInButtonText,
            signedOut, accountButtonText, signOutButtonText, accountImageURL, accountDisplayName,
        } = this.props;

        return (
            <div className="navigation__account">
                {signedOut && (
                    <Button
                        appearance="secondary"
                        className="navigation__account-sign-in-btn"
                        onClickHandler={onSignInClickHandler}
                    >
                        {signInButtonText}
                    </Button>
                )}
                {!signedOut && (
                    <SubNavigationMobileMenu
                        menuLabel={(
                            <div className="layout layout--align-middle">
                                <div className="layout__item u-fit">
                                    <ProfileImage imageURL={accountImageURL} size="3.6rem" />
                                </div>
                                <div className="layout__item u-fill">
                                    <div className="navigation__account-name truncate">{accountDisplayName}</div>
                                </div>
                            </div>
                        )}
                    >
                        <ul className="sub-navigation-mobile-menu__options drop-down-menu-base__menu-list">
                            {manageAccountURL && (
                                <li>
                                    <a href={manageAccountURL} className="drop-down-menu-base__item">
                                        <span className="label">{accountButtonText}</span>
                                    </a>
                                </li>
                            )}

                            {onSignOutClickHandler && (
                                <li>
                                    <Button className="drop-down-menu-base__item" onClickHandler={onSignOutClickHandler} appearance="no-style">
                                        <span className="label">{signOutButtonText}</span>
                                    </Button>
                                </li>
                            )}
                        </ul>
                    </SubNavigationMobileMenu>
                )}
            </div>
        );
    }

    private _renderNavLinks(): JSX.Element {
        const { visibleNavLinks } = this.props;

        return (
            <ul className="drop-down-menu-base__menu-list">
                {visibleNavLinks.map(navLink => (
                    <li key={navLink.path}>
                        {!navLink.subRoutes && (
                            <NavLink
                                to={navLink.path}
                                activeClassName="active"
                                className="drop-down-menu-base__item"
                                onClick={this._onCloseMobileMenuHandler}
                            >
                                <span className="label">{navLink.label}</span>
                            </NavLink>
                        )}
                        {navLink.subRoutes && (
                            <SubNavigationMobileMenu
                                menuLabel={navLink.label}
                            >
                                <ul className="sub-navigation-mobile-menu__options drop-down-menu-base__menu-list">
                                    {navLink.subRoutes && navLink.subRoutes.map(subNavItem => (
                                        <li key={subNavItem.path}>
                                            <NavLink
                                                key={subNavItem.path}
                                                to={subNavItem.path}
                                                activeClassName="active"
                                                className="drop-down-menu-base__item"
                                                onClick={this._onCloseMobileMenuHandler}
                                            >
                                                <span className="label">{subNavItem.label}</span>
                                            </NavLink>
                                        </li>
                                    ))}
                                </ul>
                            </SubNavigationMobileMenu>
                        )}
                    </li>

                ))}
            </ul>
        );
    }

    render(): JSX.Element {
        const { showMobileMenu, showMobileAccountNav } = this.props;

        return (
            <div className="navigation__mobile hide-sm">
                <SlideInPanel
                    isOpen={showMobileMenu}
                    width="30rem"
                    includeHeader={false}
                    onOverlayClickHandler={this._onCloseMobileMenuHandler}
                    slideDirection="left"
                >
                    <div>
                        {showMobileAccountNav && this._renderAccountNav()}
                        {this._renderNavLinks()}
                    </div>
                </SlideInPanel>
            </div>
        );
    }
}

export default NavigationMobile;
