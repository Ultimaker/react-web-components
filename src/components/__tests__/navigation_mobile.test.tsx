// Copyright (c) 2018 Ultimaker B.V.
import * as React from 'react';
import { shallow } from 'enzyme';
import { BrowserRouter, NavLink } from 'react-router-dom';

// component
import { NavigationMobile, NavigationMobileProps } from '../navigation_mobile';

describe('The NavigationMobile component', () => {
    let props: NavigationMobileProps;
    let wrapper;

    beforeEach(() => {
        props = {
            visibleNavLinks: [
                {
                    path: '/home',
                    label: 'Home',
                    visible: true,
                },
            ],
            accountButtonText: 'Account',
            signOutButtonText: 'Sign out',
            onCloseMobileMenuHandler: jest.fn(),
            showMobileAccountNav: true,
        };
        wrapper = shallow(<NavigationMobile {...props} />);
    });

    it('should render', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should render sub navigation', () => {
        wrapper.setProps({
            visibleNavLinks: [
                {
                    path: '/home',
                    label: 'Home',
                    visible: true,
                    subRoutes: [
                        {
                            path: '/home/teams',
                            label: 'Teams',
                        },
                        {
                            path: '/home/printers',
                            label: 'Printers',
                        },
                    ],
                },
            ],
        });
        expect(wrapper).toMatchSnapshot();
    });

    it('should close mobile navigation menu when navigating', () => {
        wrapper.find(NavLink).first().props().onClick();
        expect(props.onCloseMobileMenuHandler).toBeCalled();
    });

    it('should not call close handler if it is not passed', () => {
        wrapper.setProps({
            onCloseMobileMenuHandler: undefined,
        });
        wrapper.find(NavLink).first().props().onClick();
        expect(props.onCloseMobileMenuHandler).not.toBeCalled();
    });

    it('should render sign out button', () => {
        const onSignOutClickHandler = jest.fn();
        wrapper.setProps({
            onSignOutClickHandler,
        });
        expect(wrapper).toMatchSnapshot();
        wrapper.find('.navigation__account-sign-out-btn').props().onClickHandler();
        expect(onSignOutClickHandler).toBeCalled();
    });

    it('should render sign in button', () => {
        wrapper.setProps({
            signedOut: true,
        });
        expect(wrapper).toMatchSnapshot();
    });

    it('should render account link', () => {
        wrapper.setProps({
            manageAccountURL: '/signout',
        });
        expect(wrapper).toMatchSnapshot();
    });
});
