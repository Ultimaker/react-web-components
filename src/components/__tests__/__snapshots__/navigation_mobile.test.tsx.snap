// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The NavigationMobile component should render 1`] = `
<div
  className="navigation__mobile hide-sm"
>
  <SlideInPanel
    includeHeader={false}
    onOverlayClickHandler={[Function]}
    slideDirection="left"
    width="30rem"
  >
    <div>
      <div
        className="navigation__account"
      >
        <SubNavigationMobileMenu
          menuLabel={
            <div
              className="layout layout--align-middle"
            >
              <div
                className="layout__item u-fit"
              >
                <ProfileImage
                  size="3.6rem"
                />
              </div>
              <div
                className="layout__item u-fill"
              >
                <div
                  className="navigation__account-name truncate"
                />
              </div>
            </div>
          }
        >
          <ul
            className="sub-navigation-mobile-menu__options drop-down-menu-base__menu-list"
          />
        </SubNavigationMobileMenu>
      </div>
      <ul
        className="drop-down-menu-base__menu-list"
      >
        <li
          key="/home"
        >
          <NavLink
            activeClassName="active"
            aria-current="page"
            className="drop-down-menu-base__item"
            onClick={[Function]}
            to="/home"
          >
            <span
              className="label"
            >
              Home
            </span>
          </NavLink>
        </li>
      </ul>
    </div>
  </SlideInPanel>
</div>
`;

exports[`The NavigationMobile component should render account link 1`] = `
<div
  className="navigation__mobile hide-sm"
>
  <SlideInPanel
    includeHeader={false}
    onOverlayClickHandler={[Function]}
    slideDirection="left"
    width="30rem"
  >
    <div>
      <div
        className="navigation__account"
      >
        <SubNavigationMobileMenu
          menuLabel={
            <div
              className="layout layout--align-middle"
            >
              <div
                className="layout__item u-fit"
              >
                <ProfileImage
                  size="3.6rem"
                />
              </div>
              <div
                className="layout__item u-fill"
              >
                <div
                  className="navigation__account-name truncate"
                />
              </div>
            </div>
          }
        >
          <ul
            className="sub-navigation-mobile-menu__options drop-down-menu-base__menu-list"
          >
            <li>
              <a
                className="drop-down-menu-base__item"
                href="/signout"
              >
                <span
                  className="label"
                >
                  Account
                </span>
              </a>
            </li>
          </ul>
        </SubNavigationMobileMenu>
      </div>
      <ul
        className="drop-down-menu-base__menu-list"
      >
        <li
          key="/home"
        >
          <NavLink
            activeClassName="active"
            aria-current="page"
            className="drop-down-menu-base__item"
            onClick={[Function]}
            to="/home"
          >
            <span
              className="label"
            >
              Home
            </span>
          </NavLink>
        </li>
      </ul>
    </div>
  </SlideInPanel>
</div>
`;

exports[`The NavigationMobile component should render sign in button 1`] = `
<div
  className="navigation__mobile hide-sm"
>
  <SlideInPanel
    includeHeader={false}
    onOverlayClickHandler={[Function]}
    slideDirection="left"
    width="30rem"
  >
    <div>
      <div
        className="navigation__account"
      >
        <Button
          appearance="secondary"
          className="navigation__account-sign-in-btn"
          linkToNewTab={false}
          shape="rectangle"
          type="button"
        />
      </div>
      <ul
        className="drop-down-menu-base__menu-list"
      >
        <li
          key="/home"
        >
          <NavLink
            activeClassName="active"
            aria-current="page"
            className="drop-down-menu-base__item"
            onClick={[Function]}
            to="/home"
          >
            <span
              className="label"
            >
              Home
            </span>
          </NavLink>
        </li>
      </ul>
    </div>
  </SlideInPanel>
</div>
`;

exports[`The NavigationMobile component should render sign out button 1`] = `
<div
  className="navigation__mobile hide-sm"
>
  <SlideInPanel
    includeHeader={false}
    onOverlayClickHandler={[Function]}
    slideDirection="left"
    width="30rem"
  >
    <div>
      <div
        className="navigation__account"
      >
        <SubNavigationMobileMenu
          menuLabel={
            <div
              className="layout layout--align-middle"
            >
              <div
                className="layout__item u-fit"
              >
                <ProfileImage
                  size="3.6rem"
                />
              </div>
              <div
                className="layout__item u-fill"
              >
                <div
                  className="navigation__account-name truncate"
                />
              </div>
            </div>
          }
        >
          <ul
            className="sub-navigation-mobile-menu__options drop-down-menu-base__menu-list"
          >
            <li>
              <Button
                appearance="no-style"
                className="navigation__account-sign-out-btn drop-down-menu-base__item"
                linkToNewTab={false}
                onClickHandler={[MockFunction]}
                shape="rectangle"
                type="button"
              >
                <span
                  className="label"
                >
                  Sign out
                </span>
              </Button>
            </li>
          </ul>
        </SubNavigationMobileMenu>
      </div>
      <ul
        className="drop-down-menu-base__menu-list"
      >
        <li
          key="/home"
        >
          <NavLink
            activeClassName="active"
            aria-current="page"
            className="drop-down-menu-base__item"
            onClick={[Function]}
            to="/home"
          >
            <span
              className="label"
            >
              Home
            </span>
          </NavLink>
        </li>
      </ul>
    </div>
  </SlideInPanel>
</div>
`;

exports[`The NavigationMobile component should render sub navigation 1`] = `
<div
  className="navigation__mobile hide-sm"
>
  <SlideInPanel
    includeHeader={false}
    onOverlayClickHandler={[Function]}
    slideDirection="left"
    width="30rem"
  >
    <div>
      <div
        className="navigation__account"
      >
        <SubNavigationMobileMenu
          menuLabel={
            <div
              className="layout layout--align-middle"
            >
              <div
                className="layout__item u-fit"
              >
                <ProfileImage
                  size="3.6rem"
                />
              </div>
              <div
                className="layout__item u-fill"
              >
                <div
                  className="navigation__account-name truncate"
                />
              </div>
            </div>
          }
        >
          <ul
            className="sub-navigation-mobile-menu__options drop-down-menu-base__menu-list"
          />
        </SubNavigationMobileMenu>
      </div>
      <ul
        className="drop-down-menu-base__menu-list"
      >
        <li
          key="/home"
        >
          <SubNavigationMobileMenu
            menuLabel="Home"
          >
            <ul
              className="sub-navigation-mobile-menu__options drop-down-menu-base__menu-list"
            >
              <li
                key="/home/teams"
              >
                <NavLink
                  activeClassName="active"
                  aria-current="page"
                  className="drop-down-menu-base__item"
                  key="/home/teams"
                  onClick={[Function]}
                  to="/home/teams"
                >
                  <span
                    className="label"
                  >
                    Teams
                  </span>
                </NavLink>
              </li>
              <li
                key="/home/printers"
              >
                <NavLink
                  activeClassName="active"
                  aria-current="page"
                  className="drop-down-menu-base__item"
                  key="/home/printers"
                  onClick={[Function]}
                  to="/home/printers"
                >
                  <span
                    className="label"
                  >
                    Printers
                  </span>
                </NavLink>
              </li>
            </ul>
          </SubNavigationMobileMenu>
        </li>
      </ul>
    </div>
  </SlideInPanel>
</div>
`;
